### PR TITLE
Print multiple node roles, remove kubeadm-specific annotation from kubectl

### PR DESCRIPTION
**Pull request by liggitt**
related to #50010

Follow up to https://github.com/kubernetes/kubernetes/pull/50438 that removes the kubeadm-specific label, makes kubectl role-agnostic, and outputs multiple roles if present

### CODE REVIEW

File: pkg/printers/internalversion/printers.go [hunk 1]
```diff
@@ -1201,22 +1196,24 @@ func getNodeExternalIP(node *api.Node) string {
...
-		return "Master"
-	}
-	if role := node.Labels[NodeLabelRole]; role != "" {
-		return strings.Title(role)
-	}
-	if role := node.Labels[NodeLabelKubeadmAlphaRole]; role != "" {
-		return strings.Title(role)
+// findNodeRoles returns the roles of a given node.
+// The roles are determined by looking for:
+// * a node-role.kubernetes.io/<role>="" label
```
Comment (dixudx) [inline pkg/printers/internalversion/printers.go]: Seems we've only got `node-role.kubernetes.io/master` for now.

File: pkg/printers/internalversion/printers_test.go [hunk 2]
```diff
@@ -829,29 +829,19 @@ func TestPrintNodeStatus(t *testing.T) {
 			node: api.Node{
 				ObjectMeta: metav1.ObjectMeta{
 					Name:   "foo10",
-					Labels: map[string]string{"node-role.kubernetes.io/master": "", "kubernetes.io/role": "node", "kubeadm.alpha.kubernetes.io/role": "node"},
+					Labels: map[string]string{"node-role.kubernetes.io/master": "", "node-role.kubernetes.io/proxy": "", "kubernetes.io/role": "node"},
 				},
 			},
-			status: "Unknown,Master",
+			status: "master,node,proxy",
```
Comment (dixudx) [inline pkg/printers/internalversion/printers_test.go]: Since we have added a new column named `"Role"`, these testcases should be moved to a new test `TestPrintNodeRole`. They do not belong to `Status` any more.

File: pkg/printers/internalversion/printers.go (hunk 1 above)
Comment (liggitt) [inline pkg/printers/internalversion/printers.go]: The point of that format was to support multiple roles, and to allow things like this to treat roles generically

File: pkg/printers/internalversion/printers.go [hunk 3]
```diff
@@ -62,16 +62,12 @@ import (
 const (
 	loadBalancerWidth = 16
 
-	// LabelNodeRoleMaster specifies that a node is a master
+	// labelNodeRolePrefix is a label prefix for node roles
 	// It's copied over to here until it's merged in core: https://github.com/kubernetes/kubernetes/pull/39112
```
Comment (luxas) [inline pkg/printers/internalversion/printers.go]: You can remove this now; as this is only the prefix; not the well-known value for the master

File: pkg/printers/internalversion/describe.go [hunk 4]
```diff
@@ -2436,7 +2436,9 @@ func describeNode(node *api.Node, nodeNonTerminatedPodsList *api.PodList, events
 	return tabbedString(func(out io.Writer) error {
 		w := NewPrefixWriter(out)
 		w.Write(LEVEL_0, "Name:\t%s\n", node.Name)
-		w.Write(LEVEL_0, "Role:\t%s\n", findNodeRole(node))
+		if roles := findNodeRoles(node); len(roles) > 0 {
+			w.Write(LEVEL_0, "Role:\t%s\n", strings.Join(roles, ","))
```
Comment (luxas) [inline pkg/printers/internalversion/describe.go]: nit: `Roles:`?

File: pkg/printers/internalversion/printers_test.go [hunk 5]
```diff
@@ -829,29 +829,19 @@ func TestPrintNodeStatus(t *testing.T) {
...
 		},
 		{
 			node: api.Node{
```
Comment (luxas) [inline pkg/printers/internalversion/printers_test.go]: Add a test case for none of these labels?

File: pkg/printers/internalversion/printers_test.go (hunk 2 above)
Comment (dixudx) [inline pkg/printers/internalversion/printers_test.go]: @liggitt Seems `TestPrintNodeStatus` did not find out the bug that `Role` is not represented with `Status`.  

```Golang
if !contains(strings.Fields(buffer.String()), test.status) {
	t.Fatalf("Expect printing node %s with status %#v, got: %#v", test.node.Name, test.status, buffer.String())
}
```

This check is not strictly matched, shall we use `strings.Fields(buffer.String())[2]` instead of this. Of course we will make sure there is no index out of range.

If there is a consensus, I would like to submit a separate PR fixing this.

File: pkg/printers/internalversion/printers.go [hunk 6]
```diff
@@ -1167,12 +1164,10 @@ func printNode(obj *api.Node, options printers.PrintOptions) ([]metav1alpha1.Tab
...
 	}
-	role := findNodeRole(obj)
-	if role != "" {
-		status = append(status, role)
-	}
 
-	row.Cells = append(row.Cells, obj.Name, strings.Join(status, ","), translateTimestamp(obj.CreationTimestamp), obj.Status.NodeInfo.KubeletVersion)
+	roles := findNodeRoles(obj)
+
+	row.Cells = append(row.Cells, obj.Name, strings.Join(status, ","), strings.Join(roles, ","), translateTimestamp(obj.CreationTimestamp), obj.Status.NodeInfo.KubeletVersion)
```
Comment (luxas) [inline pkg/printers/internalversion/printers.go]: Should we still print this column if `len(roles) == 0`? I have no specific opinion on that...
Or should we default to just print `node`?
(just throwing out ideas here)

**Review by luxas — COMMENTED** [review]
Agree with the earlier comments too, left two nits and some questions.
The unit test case should probably be cleaned up to account for the new Role field, though.

Thanks for removing the old kubeadm label again :)
(Tried to do it in https://github.com/kubernetes/kubernetes/pull/48056 already, but it was reasonable to revert those to get us back to this earlier state :+1:)

File: pkg/printers/internalversion/printers.go (hunk 6 above)
Comment (dixudx) [inline pkg/printers/internalversion/printers.go]: +1. Same concern.

> Or should we default to just print node?

For mixed case, we'd better have a default value.

```Golang
if len(roles) == 0 {
    roles = append(roles, "<unknown>")
    // or roles = append(roles, "Unknown")
}
```

WDYT?

File: pkg/printers/internalversion/printers_test.go (hunk 2 above)
Comment (liggitt) [inline pkg/printers/internalversion/printers_test.go]: > shall we use strings.Fields(buffer.String())[2] instead of this

a tighter check could be ok, but I'd rather not make our tests brittle to adding or removing unrelated columns. follow up is fine

File: pkg/printers/internalversion/printers.go (hunk 3 above)
Comment (liggitt) [inline pkg/printers/internalversion/printers.go]: I'd expect the prefix constant to be made part of the linked PR

File: pkg/printers/internalversion/printers.go (hunk 6 above)
Comment (liggitt) [inline pkg/printers/internalversion/printers.go]: defaulted to `<none>` if unlabeled

**Comment by liggitt** [comment]
comments addressed

**Comment by liggitt** [comment]
W0823 13:37:32.109] ERROR: error loading package '': Encountered error while reading extension file 'go/def.bzl': no such package '@io_bazel_rules_go//go': Error downloading [https://github.com/bazelbuild/rules_go/archive/82483596ec203eb9c1849937636f4cbed83733eb.tar.gz] to /root/.cache/bazel/_bazel_root/e9f728bbd90b3fba632eb31b20e1dacd/external/io_bazel_rules_go

/retest

**Review by luxas — APPROVED** [review]
/lgtm
